### PR TITLE
px4fmu move RC input to new rc_input driver

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

**Comment by dagar** [comment]
This saves about a half kilobyte of flash on fmu-v2.

**Comment by dagar** [comment]
Tested on a pixracer (both bind and regular usage).

**Comment by mhkabir** [comment]
What is the effect of this on latency? Curious how this will affect racers.

**Comment by dagar** [comment]
I was a slightly concerned with that as well, which is why I have it running at 200Hz. We could also add some logic for the work queue delay that syncs up with RC serial decoding after a successful parse.

If I can save enough memory through other means I would come back to this and have it run as a task.

Also keep in mind that it wasn't that long ago the pixracer was only running fmu in the work queue. https://github.com/PX4/Firmware/pull/9388

**Comment by bkueng** [comment]
> which is why I have it running at 200Hz

This PR has it set to 100Hz. I'm concerned about the delay too, and RC delay was one of the things I wanted to look into as well. Can you extend it to make it possible to run it as task like fmu?

While this is a good cleanup and improves modularity, we have to be careful not to have too many small different tasks. Each comes with scheduling & memory overhead.
Did you compare RAM usage?

**Comment by dagar** [comment]
Thanks, looks like I didn't push the change to 200 Hz.

This PR is an extra 80 bytes of RAM. No noticeable cpu difference.

Stripping the px4fmu module down to be an output module and not a random catchall is a small part of getting the system ready to run the rate controller and output at 400 Hz - 1 kHz with as little latency and jitter as possible.

**Comment by dagar** [comment]
Updated to run at 250Hz, so for now it's effectively the same as how it currently runs in px4fmu.

**Comment by dagar** [comment]
> Can you extend it to make it possible to run it as task like fmu?

Yes, but I wasn't planning on doing it here unless someone would like to help with testing. We'd need to change the structure a bit to optionally poll the serial port and make sure all the scanning logic still makes sense. I'd also like to merge with the spektrum_rc driver (https://github.com/PX4/Firmware/blob/master/src/drivers/spektrum_rc/spektrum_rc.cpp).

**Comment by bkueng** [comment]
> Yes, but I wasn't planning on doing it here unless someone would like to help with testing.

I can help with that.

**Comment by dagar** [comment]
I did the split again with the current state of the px4fmu driver. This time I left the task support in.

I'll do some basic testing today and if there's nothing obvious I'd like to get this merged before it rots again.

**Comment by dagar** [comment]
TODO: merge with src/drivers/spektrum_rc (at a later date)

**Comment by dagar** [comment]
Can we find a better way to handle the CONFIG_ARCH_BOARD_OMNIBUS_F4SD define for CRSFTelemetry?

**Comment by dagar** [comment]
TODO: proper polling logic

**Comment by dagar** [comment]
I've tested spektrum on a pixracer.

File: src/drivers/px4fmu/fmu.cpp [hunk 1]
```diff
@@ -1228,7 +1033,7 @@ PX4FMU::cycle()
 			_current_update_rate = 0;
 		}
 
-		int poll_timeout = 5; // needs to be small enough so that we don't miss RC input data
+		int poll_timeout = 5;
```
Comment (bkueng) [inline src/drivers/px4fmu/fmu.cpp]: We can increase this to 10 or so

File: src/drivers/rc_input/RCInput.hpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+#include <uORB/topics/input_rc.h>
+#include <uORB/topics/vehicle_command.h>
+
+#include "crsf_telemetry.h"
+
+#ifdef HRT_PPM_CHANNEL
+# include <systemlib/ppm_decode.h>
+#endif
+
+enum RC_SCAN {
```
Comment (bkueng) [inline src/drivers/rc_input/RCInput.hpp]: Can you move this and RC_SCAN_STRING into the class?

File: src/drivers/rc_input/RCInput.cpp [hunk 3]
```diff
@@ -0,0 +1,816 @@
...
+{
+	_adc_sub = orb_subscribe(ORB_ID(adc_report));
+
+#ifdef RC_SERIAL_PORT
+
+#  ifdef RF_RADIO_POWER_CONTROL
+	// power radio on
+	RF_RADIO_POWER_CONTROL(true);
+#  endif
+	_vehicle_cmd_sub = orb_subscribe(ORB_ID(vehicle_command));
```
Comment (bkueng) [inline src/drivers/rc_input/RCInput.cpp]: This is never unsubscribed

File: src/drivers/rc_input/RCInput.cpp [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+					      (px4_main_t)&run_trampoline,
+					      nullptr);
+
+		if (_task_id < 0) {
+			_task_id = -1;
+			return -errno;
+		}
+	}
+
+	// wait until task is up & running (the mode_* commands depend on it)
```
Comment (bkueng) [inline src/drivers/rc_input/RCInput.cpp]: left-over comment. I don't think we need to wait here.

File: src/drivers/px4fmu/fmu.cpp (hunk 1 above)
Comment (dagar) [reply src/drivers/px4fmu/fmu.cpp]: Done

File: src/drivers/rc_input/RCInput.hpp (hunk 2 above)
Comment (dagar) [reply src/drivers/rc_input/RCInput.hpp]: Done

File: src/drivers/rc_input/RCInput.cpp (hunk 3 above)
Comment (dagar) [reply src/drivers/rc_input/RCInput.cpp]: Fixed

File: src/drivers/rc_input/RCInput.cpp (hunk 4 above)
Comment (dagar) [reply src/drivers/rc_input/RCInput.cpp]: Fixed

Comment (dagar) [reply src/drivers/rc_input/RCInput.cpp]: Dropped

File: src/drivers/rc_input/RCInput.cpp [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+
+#ifdef RC_SERIAL_PORT
+
+#  ifdef RF_RADIO_POWER_CONTROL
+	// power radio on
+	RF_RADIO_POWER_CONTROL(true);
+#  endif
+	_vehicle_cmd_sub = orb_subscribe(ORB_ID(vehicle_command));
+	// dsm_init sets some file static variables and returns a file descriptor
+	_rcs_fd = dsm_init(RC_SERIAL_PORT);
```
Comment (bkueng) [inline src/drivers/rc_input/RCInput.cpp]: `_rcs_fd` is never closed.
The rest should be good, I'm ok implementing the polling in the next iteration.

Comment (bkueng) [reply src/drivers/rc_input/RCInput.cpp]: > _rcs_fd is never closed.

Never mind, `dsm_deinit` does it.

**Pull request merged**